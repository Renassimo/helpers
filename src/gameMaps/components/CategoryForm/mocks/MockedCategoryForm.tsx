const MockedCategoryForm = jest.fn(({ values, errors, isReady }) => (
  <div>
    MockedCategoryForm - {values?.title} - {values?.description} -{' '}
    {values?.color} - {values?.itemsAmount} - {errors?.title} -{' '}
    {errors?.description} - {errors?.color} - {errors?.itemsAmount} -{' '}
    {isReady && 'isReady'}
  </div>
));

export default MockedCategoryForm;
