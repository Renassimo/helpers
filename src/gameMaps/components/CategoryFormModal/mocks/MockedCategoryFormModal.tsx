const MockedCategoryFormModal = jest.fn(({ isModalOpen, gameId }) => (
  <div>
    MockedCategoryFormModal {isModalOpen ? 'opened' : 'closed'} - {gameId}
  </div>
));

export default MockedCategoryFormModal;
