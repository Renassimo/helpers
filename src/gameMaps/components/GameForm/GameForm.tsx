import TextField from '@mui/material/TextField';
import Alert from '@mui/material/Alert';

import ImagePicker from '@/common/components/ImagePicker';
import ColorPicker from '@/common/components/ColorPicker';
import Button from '@mui/material/Button';

import { GameFormProps } from '@/gameMaps/types/props';

const GameForm = ({
  values,
  setters,
  errors = {},
  onDelete,
}: GameFormProps) => {
  const { title, description, backgroundColor, mapImageUrl } = values;
  const { setTitle, setDescription, setBackgroundColor, setMapImage } = setters;

  return (
    <>
      <TextField
        name="title"
        label="Title"
        value={title}
        onChange={(event) => setTitle(event.target.value)}
        fullWidth
        margin="dense"
        variant="standard"
        error={!!errors.title}
        helperText={errors.title}
      />
      <TextField
        name="description"
        label="Description"
        value={description}
        onChange={(event) => setDescription(event.target.value)}
        fullWidth
        multiline
        margin="dense"
        variant="standard"
        error={!!errors.description}
        helperText={errors.description}
      />
      <ColorPicker
        name="backgroundColor"
        label="Map background color"
        value={backgroundColor}
        onChange={setBackgroundColor}
        error={errors.backgroundColor}
      />
      <ImagePicker defaultUrlValue={mapImageUrl} onChange={setMapImage} />
      {onDelete && (
        <Button
          type="button"
          fullWidth
          variant="outlined"
          onClick={onDelete}
          color="error"
        >
          Delete
        </Button>
      )}
      {errors.main && <Alert severity="error">{errors.main}</Alert>}
    </>
  );
};

export default GameForm;
