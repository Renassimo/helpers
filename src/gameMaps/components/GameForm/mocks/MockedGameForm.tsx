const MockedGameForm = jest.fn(({ values, errors }) => (
  <div>
    MockedGameForm - {values?.title} - {values?.description} -{' '}
    {values?.backgroundColor} - {values?.mapImageUrl} - {errors?.title} -{' '}
    {errors?.description} - {errors?.backgroundColor}
  </div>
));

export default MockedGameForm;
