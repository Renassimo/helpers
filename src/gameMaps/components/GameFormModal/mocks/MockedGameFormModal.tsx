const MockedGameFormModal = jest.fn(({ isModalOpen }) => (
  <div>MockedGameFormModal {isModalOpen ? 'opened' : 'closed'}</div>
));

export default MockedGameFormModal;
