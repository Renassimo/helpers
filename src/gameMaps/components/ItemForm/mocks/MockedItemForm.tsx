const MockedItemForm = jest.fn(({ values, errors }) => (
  <div>
    MockedItemForm - {values?.description} - {values?.collected && 'collected'}{' '}
    - {errors?.description} - {errors?.collected}
  </div>
));

export default MockedItemForm;
