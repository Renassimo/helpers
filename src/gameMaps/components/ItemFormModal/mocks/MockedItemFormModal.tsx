const MockedItemFormModal = jest.fn(
  ({ isModalOpen, gameId, coordinates, playId, categoryId }) => (
    <div>
      MockedItemFormModal {isModalOpen ? 'opened' : 'closed'} - {gameId} -{' '}
      {coordinates?.[0]} - {coordinates?.[1]} - {playId} - {categoryId}
    </div>
  )
);

export default MockedItemFormModal;
