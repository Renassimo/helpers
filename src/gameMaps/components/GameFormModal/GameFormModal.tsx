import { useEffect } from 'react';

import useGameForm from '@/gameMaps/hooks/useGameForm';

import Modal from '@/common/components/Modal';
import GameForm from '@/gameMaps/components/GameForm';

import { GameData } from '@/gameMaps/types';

const GameFormModal = ({
  isModalOpen,
  setIsModalOpen,
  data,
  onFinish,
}: {
  isModalOpen: boolean;
  setIsModalOpen: (isOpen: boolean) => void;
  data?: GameData | null;
  onFinish: (newData: GameData | null) => void;
}) => {
  const onModalClose = () => {
    setIsModalOpen(false);
    cleanForm();
  };

  const handleFinish = (newData: GameData | null) => {
    onFinish(newData);
    onModalClose();
  };

  const {
    isEditForm,
    values,
    setters,
    cleanForm,
    prepareFormForEdit,
    onSubmit,
    onDelete,
    errors,
    loading,
  } = useGameForm(data, handleFinish);

  const handleDelete = async (): Promise<void> => {
    if (
      confirm(
        `Are you sure to delete "${
          data?.attributes.title ?? values.title
        }" game? Following action will also delete all plays, categories, all found items releated to this game and map image.`
      )
    ) {
      await onDelete();
      handleFinish(null);
    }
  };

  useEffect(() => {
    if (isModalOpen && isEditForm) {
      prepareFormForEdit();
    }
  }, [isModalOpen, isEditForm]);

  return (
    <Modal
      open={isModalOpen}
      onClose={onModalClose}
      onSubmit={onSubmit}
      title={`${data ? 'Create new' : `Update`} game`}
      loading={loading}
    >
      <GameForm
        values={values}
        setters={setters}
        errors={errors}
        onDelete={isEditForm ? handleDelete : undefined}
      />
    </Modal>
  );
};

export default GameFormModal;
