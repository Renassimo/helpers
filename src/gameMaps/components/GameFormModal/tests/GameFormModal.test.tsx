import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import useGameForm from '@/gameMaps/hooks/useGameForm';

import Modal from '@/common/components/Modal';
import GameForm from '@/gameMaps/components/GameForm';

import MockedModal from '@/common/components/Modal/mocks';
import MockedGameForm from '@/gameMaps/components/GameForm/mocks';
import { mockedGame } from '@/gameMaps/types/mocks';

import GameFormModal from '../GameFormModal';

jest.mock('@/gameMaps/hooks/useGameForm');
jest.mock('@/common/components/Modal');
jest.mock('@/gameMaps/components/GameForm');

describe('GameFormModal', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    (Modal as unknown as jest.Mock).mockImplementation(MockedModal);
    (GameForm as unknown as jest.Mock).mockImplementation(MockedGameForm);
  });

  const mockedData = mockedGame;

  const mockedValues = {
    title: 'mocked-title',
    description: 'mocked-description',
    backgroundColor: 'mocked-background-color',
    mapImageUrl: 'mocked-map-image-url',
  };
  const mockedSetters = 'mocked-setters';
  const mockedCleanForm = jest.fn();
  const mockedPrepareFormForEdit = jest.fn();
  const mockedOnSubmit = jest.fn();
  const mockedSetIsModalOpen = jest.fn();
  const mockedErrors = 'mockedErrors';
  const mockedOnFinish = jest.fn();

  test('renders snapshot successfully', () => {
    // Arange
    const mockedUseGameForm = jest.fn(() => ({
      isEditForm: false,
      values: mockedValues,
      setters: mockedSetters,
      cleanForm: mockedCleanForm,
      prepareFormForEdit: mockedPrepareFormForEdit,
      onSubmit: mockedOnSubmit,
      errors: mockedErrors,
    }));
    (useGameForm as unknown as jest.Mock).mockImplementation(mockedUseGameForm);
    // Act
    const { baseElement } = renderWithTheme(
      <GameFormModal
        isModalOpen={true}
        setIsModalOpen={mockedSetIsModalOpen}
        onFinish={mockedOnFinish}
      />
    );
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(mockedUseGameForm).toBeCalledWith(undefined, expect.any(Function));
    expect(mockedPrepareFormForEdit).not.toBeCalled();
    expect(MockedGameForm).toHaveBeenCalledWith(
      { values: mockedValues, setters: mockedSetters, errors: mockedErrors },
      {}
    );
  });

  describe('when data passed', () => {
    test('calls prepareFormForEdit', () => {
      // Arange
      const mockedUseGameForm = jest.fn(() => ({
        isEditForm: true,
        values: mockedValues,
        setters: mockedSetters,
        cleanForm: mockedCleanForm,
        prepareFormForEdit: mockedPrepareFormForEdit,
        onSubmit: mockedOnSubmit,
        errors: mockedErrors,
      }));
      (useGameForm as unknown as jest.Mock).mockImplementation(
        mockedUseGameForm
      );
      // Act
      renderWithTheme(
        <GameFormModal
          isModalOpen={true}
          setIsModalOpen={mockedSetIsModalOpen}
          data={mockedData}
          onFinish={mockedOnFinish}
        />
      );
      // Assert
      expect(mockedUseGameForm).toBeCalledWith(
        mockedData,
        expect.any(Function)
      );
      expect(mockedPrepareFormForEdit).toBeCalled();
    });
  });
});
