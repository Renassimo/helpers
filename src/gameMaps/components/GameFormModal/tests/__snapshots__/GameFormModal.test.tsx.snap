// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameFormModal renders snapshot successfully 1`] = `
<body>
  <div>
    <div>
      <div>
        MockedGameForm - 
        mocked-title
         - 
        mocked-description
         -
         
        mocked-background-color
         - 
        mocked-map-image-url
         - 
         -
         
         - 
      </div>
      <div>
        title: 
        Update game
        ; maxWidth: 
        sm
        ;
        open
        ;
        not loading
        ;
        not disabled
        ; onClose
         
        passed
        ; onSubmit
         
        passed
        ;
      </div>
    </div>
  </div>
</body>
`;

exports[`GameFormModal when data passed renders snapshot successfully 1`] = `
<body>
  <div>
    <div>
      <div>
        MockedGameForm - 
        mocked-title
         - 
        mocked-description
         -
         
        mocked-background-color
         - 
        mocked-map-image-url
         - 
         -
         
         - 
      </div>
      <div>
        title: 
        Create new game
        ; maxWidth: 
        sm
        ;
        open
        ;
        not loading
        ;
        not disabled
        ; onClose
         
        passed
        ; onSubmit
         
        passed
        ;
      </div>
    </div>
  </div>
</body>
`;
